### PR TITLE
find_if conversion

### PR DESCRIPTION
**Pull request by neheb**
Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

File: src/KaxCluster.cpp [hunk 1]
```diff
@@ -180,10 +179,13 @@ filepos_t KaxCluster::Render(IOCallback & output, KaxCues & CueToUpdate, bool bS
   } else {
     // new school, using KaxBlockBlob
     for (const auto& blob : Blobs) {
-      if (blob->IsSimpleBlock())
-        PushElement( static_cast<KaxSimpleBlock&>(*blob));
-      else
-        PushElement( static_cast<KaxBlockGroup&>(*blob));
+      if (blob->IsSimpleBlock()) {
+        auto simple = KaxSimpleBlock(*blob);
+        PushElement(simple);
```
Comment (mbunkus) [inline src/KaxCluster.cpp:184]: This is wrong. The original code simply passes a memory address along (in form of a reference), while your change will make copies of objects.

File: src/KaxCluster.cpp [hunk 2]
```diff
@@ -194,12 +196,10 @@ filepos_t KaxCluster::Render(IOCallback & output, KaxCues & CueToUpdate, bool bS
...
           auto tracknum = static_cast<std::uint32_t>(entry->TrackNumber());
-          for (Index = 0; Index<Blobs.size(); Index++) {
-            if (static_cast<KaxInternalBlock&>(*Blobs[Index]).TrackNum() == tracknum)
-              break; // this track is used
-          }
+          auto it = std::find_if(Blobs.begin(), Blobs.end(), [=](auto b){ return static_cast<const KaxInternalBlock&>(*b).TrackNum() == tracknum; });
+
           // the track wasn't found in this cluster
-          if (Index == ListSize()) {
+          if (static_cast<std::size_t>(std::distance(Blobs.begin(), it)) == ListSize()) {
```
Comment (mbunkus) [inline src/KaxCluster.cpp:202]: The more idiomatic and easier to understand way to write this is `if (it == Blobs.end()) {`.

File: src/KaxCluster.cpp [hunk 3]
```diff
@@ -194,12 +196,10 @@ filepos_t KaxCluster::Render(IOCallback & output, KaxCues & CueToUpdate, bool bS
         if (EbmlId(*Trk) == EBML_ID(KaxTrackEntry)) {
           auto entry = static_cast<KaxTrackEntry *>(Trk);
           auto tracknum = static_cast<std::uint32_t>(entry->TrackNumber());
-          for (Index = 0; Index<Blobs.size(); Index++) {
-            if (static_cast<KaxInternalBlock&>(*Blobs[Index]).TrackNum() == tracknum)
-              break; // this track is used
-          }
+          auto it = std::find_if(Blobs.begin(), Blobs.end(), [=](auto b){ return static_cast<const KaxInternalBlock&>(*b).TrackNum() == tracknum; });
```
Comment (mbunkus) [inline src/KaxCluster.cpp:199]: Please name the values to copy into the lamba explicitly, e.g. `…, [=tracknum](auto b){ …` That's much more explicit, and it'll never lead to accidentally making copies of objects one might not have wanted to copy if there are ever changes to the lambda's code.

**Review by mbunkus — CHANGES_REQUESTED** [review]

File: src/KaxCluster.cpp (hunk 2 above)
Comment (neheb) [reply src/KaxCluster.cpp:202]: I wasn't sure if ListSize() is equivalent.

**Comment by neheb** [comment]
Removed first commit.

File: src/KaxCluster.cpp (hunk 2 above)
Comment (mbunkus) [reply src/KaxCluster.cpp]: The longer I look at this code the more I'm convinced the original code was actually buggy. The function `ListSize()` is only defined in the `EbmlMaster` class and never overridden. It simply returns `ElementList.size()`.

However, this piece of code looks through a different vector, `Blobs`, a member of the `KaxCluster` class.

The code should probably have read `if (Index == Blobs.ListSize()) {` before, making `if (it == Blobs.end()) {` even more correct than I initially thought.

Comment (mbunkus) [reply src/KaxCluster.cpp]: The reason why no one ever noticed is that the code is only run if `bSilentTracksUsed` is `true`, and I don't think there's any program that actually uses this "silent tracks" feature…

Comment (neheb) [reply src/KaxCluster.cpp]: I noticed there are a bunch of files in test/ that don't work. They need updating it seems.

There are even unicode issues in some of them that prevents GCC from compiling.

**Comment by mbunkus** [comment]
Thanks.

**Pull request merged**